pkg/resource/v1/certconfig/desired.go

```diff
@@ -2,8 +2,375 @@ package certconfig
 
 import (
 	"context"
+	"strings"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/certs"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/versionbundle"
+	apimetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/cluster-operator/pkg/cluster"
+	"github.com/giantswarm/cluster-operator/pkg/label"
+	"github.com/giantswarm/cluster-operator/pkg/resource/v1/certconfig/key"
+)
+
+const (
+	certAPIVersion = "core.giantswarm.io"
+	certKind       = "CertConfig"
+	certOperatorID = "cert-operator"
+
+	loopbackIP = "127.0.0.1"
+
+	// systemMastersOrganization is the RBAC kubernetes admin group.
+	systemMastersOrganization = "system:masters"
+)
+
+var (
+	kubeAltNames = []string{
+		"kubernetes",
+		"kubernetes.default",
+		"kubernetes.default.svc",
+		"kubernetes.default.svc.cluster.local",
+	}
 )
 
+// GetDesiredState returns all desired CertConfigs for managed certificates.
 func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
-	return nil, nil
+	clusterGuestConfig, err := r.toClusterGuestConfigFunc(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	clusterConfig, err := prepareClusterConfig(*r.baseClusterConfig, *clusterGuestConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredCertConfigs := make([]*v1alpha1.CertConfig, 0)
+	{
+		certConfig := newAPICertConfig(clusterConfig, certs.APICert, r.projectName)
+		desiredCertConfigs = append(desiredCertConfigs, certConfig)
+	}
+	{
+		certConfig := newCalicoCertConfig(clusterConfig, certs.CalicoCert, r.projectName)
+		desiredCertConfigs = append(desiredCertConfigs, certConfig)
+	}
+	{
+		certConfig := newEtcdCertConfig(clusterConfig, certs.EtcdCert, r.projectName)
+		desiredCertConfigs = append(desiredCertConfigs, certConfig)
+	}
+	{
+		certConfig := newNodeOperatorCertConfig(clusterConfig, certs.NodeOperatorCert, r.projectName)
+		desiredCertConfigs = append(desiredCertConfigs, certConfig)
+	}
+	{
+		certConfig := newPrometheusCertConfig(clusterConfig, certs.PrometheusCert, r.projectName)
+		desiredCertConfigs = append(desiredCertConfigs, certConfig)
+	}
+	{
+		certConfig := newServiceAccountCertConfig(clusterConfig, certs.ServiceAccountCert, r.projectName)
+		desiredCertConfigs = append(desiredCertConfigs, certConfig)
+	}
+	{
+		certConfig := newWorkerCertConfig(clusterConfig, certs.WorkerCert, r.projectName)
+		desiredCertConfigs = append(desiredCertConfigs, certConfig)
+	}
+
+	return desiredCertConfigs, nil
+}
+
+func prepareClusterConfig(baseClusterConfig cluster.Config, clusterGuestConfig v1alpha1.ClusterGuestConfig) (*cluster.Config, error) {
+	var err error
+
+	// Copy baseClusterConfig as basis and supplement it with information from
+	// clusterGuestConfig.
+	clusterConfig := new(cluster.Config)
+	*clusterConfig = baseClusterConfig
```

Comment:
Plain `clusterConfig := baseClusterConfig` leads to aforementioned `return &clusterConfig` you noted earlier.

`clusterConfig = baseClusterConfig` doesn't work with current Go:
> pkg/resource/v1/certconfig/desired.go:89:16: cannot use baseClusterConfig (type cluster.Config) as type *cluster.Config in assignment

`clusterConfig = &baseClusterConfig` would be valid here, but without wider context one may not realize that `baseClusterConfig` is a copy of original here. I'd like to keep it as it currently is so that it makes it clearer that `baseClusterConfig` is copied to `clusterConfig`.